doc/namespace.md

```diff
@@ -0,0 +1,417 @@
+
+```ruby
+# foo.rb
+FOO = 100
+
+FOO         #=> 100
+Object::FOO #=> 100
+
+# main.rb
+ns = Namespace.new
+ns.require('foo')
+
+ns::FOO      #=> 100
+
+FOO          # NameError
+Object::FOO  # NameError
+```
+
+### Top level methods
+
+Top level methods are private instance methods of `Object`, in each namespace.
+
+```ruby
+# foo.rb
+def yay = "foo"
+
+class Foo
+  def self.say = yay
+end
+
+Foo.say #=> "foo"
+yay     #=> "foo"
+
+# main.rb
+ns = Namespace.new
+ns.require('foo')
+
+ns.Foo.say  #=> "foo"
+
+yay  # NoMethodError
+```
+
+There is no way to expose top level methods in namespaces to another namespace.
+(See "Expose top level methods as a method of the namespace object" in "Discussions" section below)
+
+### Namespace scopes
+
+Namespace works in file scope. One `.rb` file runs in a single namespace.
+
+Once a file is loaded in a namespace `ns`, all methods/procs defined/created in the file run in `ns`.
+
+## Implementation details
+
+#### Object Shapes
+
+Once builtin classes are copied and modified in namespaces, its instance variable management fallbacks from Object Shapes to a traditional iv table (st_table) because RClass stores the shape in its `flags`, not in `rb_classext_t`.
+
+#### Size of RClass and rb_classext_t
+
+Namespace requires to move some fields from RClass to `rb_classext_t`, then the size of RClass and `rb_classext_t` is now larger than `4 * RVALUE_SIZE`. It's against the expectation of [Variable Width Allocation](https://rubykaigi.org/2021-takeout/presentations/peterzhu2118.html).
+
+Now the `STATIC_ASSERT` to check the size is commented-out. (See "Minimize the size of RClass and rb_classext_t" in "Discussion" section below)
+
+#### ISeq inline method/constant cache
+
+As described above in "Namespace scopes", an ".rb" file runs in a namespace. So method/constant resolution will be done in a namespace consistently.
+
+That means ISeq inline caches work well even with namespaces. Otherwise, it's a bug.
+
+#### Method call global cache (gccct)
+
+`rb_funcall()` C function refers to the global cc cache table (gccct), and the cache key is calculated with the current namespace.
+
+So, `rb_funcall()` calls have a performance penalty when namespace is enabled.
+
+#### Current namespace and loading namespace
+
+The current namespace is the namespace that the executing code is in. `Namespace.current` returns the current namespace object.
+
+The loading namespace is an internally managed namespace to determine the namespace to load newly required/loaded files. For example, `ns` is the loading namespace when `ns.require("foo")` is called.
+
+## Discussions
+
+#### Namespace#inspect
+
+Currently, `Namespace#inspect` returns values like `"#<Namespace:0x00000001083a5660>"`. This results in the very redundant and poorly visible classpath outside the namespace.
```

Comment:
Not only outside the namespace but also inside, so "always":
```
$ RUBY_NAMESPACE=1 ruby -ve 'ns = Namespace.new; File.write "ns.rb", "class C; end; p C"; ns.require "./ns"'    
ruby 3.5.0dev (2025-05-10T07:50:29Z namespace-on-read-.. bd4f57f96b) +PRISM [x86_64-linux]
ruby: warning: Namespace is experimental, and the behavior may change in the future!
See doc/namespace.md for know issues, etc.
#<Namespace:0x00007fd81ce67598>::C
```

Reply:
And, the "Discussions" section is to make discussions after merging this change, and I'm not going to make comments in this pull-requests on those points.
Thank you for making the comments.